app/models/token.rb

```diff
@@ -44,6 +47,7 @@ def reassign_tokens
       self.tokens = 0
     else
       self.tokens = assignment.tokens_per_day
+      assignment.last_token_regeneration_date = DateTime.now
```

Comment:
Okay so it seems this code is called from `execute_test_run`, which can happen at arbitrary times. This means that using `DateTime.now` is not correct, since future checks will use `last_token_regeneration_date` and lose track of the scheduled regeneration times.

`assignment.last_token_generation_date` should always be equal to `assignment.tokens_start_of_availability_date + n * assignment.regeneration_period`, where `n` is some number.